loader/src/loader.c

```diff
@@ -153,9 +153,9 @@ static void
 putc(uint8_t ch)
 {
     while (!(*UART_REG(STAT) & STAT_TDRE)) { }
-    *UART_REG(TRANSMIT) = ch;
     if (ch == '\n')
         *UART_REG(TRANSMIT) = '\r';
+    *UART_REG(TRANSMIT) = ch;
```

Comment:
The UART output is much much slower than the core executing instructions. Thus you have to check here again that FIFO has space in case you wrote the `r` there. I'm not sure what the `STAT_TDRE` tells you exactly, usually it that the FIFO has space for at least one char and not that the FIFO is empty and thus can take more than one char. So better use this:
```
    if (ch == '\n') {
        // ensure FIFO has space
        while (!(*UART_REG(STAT) & STAT_TDRE)) { }
        *UART_REG(TRANSMIT) = '\r';
    }     
    
    // ensure FIFO has space
    while (!(*UART_REG(STAT) & STAT_TDRE)) { }
    *UART_REG(TRANSMIT) = ch;
```

Reply:
Thanks for this